test/compilable/b12603.d

```diff
@@ -0,0 +1,111 @@
+// PERMUTE_ARGS:
+// REQUIRED_ARGS: -o-
+
+struct S
+{
+    this(uint* dtorCalled)
+    {
+        *dtorCalled = 0;
+        this.dtorCalled = dtorCalled;
+    }
+
+    @disable this();
+
+    ~this()
+    {
+        ++*dtorCalled;
+    }
+    
+    uint* dtorCalled;
+}
+
+
+auto numDtorCallsByGotoWithinScope()
+{
+    uint dtorCalled;
+    {
+      S s = S(&dtorCalled);
+      assert(dtorCalled == 0);
+      goto L_abc;
+      L_abc:
+      assert(dtorCalled == 0);
+    }
+    assert(dtorCalled == 1);
+    return dtorCalled;
+}
+static assert(numDtorCallsByGotoWithinScope() == 1);
+
+
+auto numDtorCallsByGotoOutOfScope()
+{
+    uint dtorCalled;
+    {
+      S s = S(&dtorCalled);
+      assert(dtorCalled == 0);
+      goto L_abc;
+    }
+    L_abc:
+    assert(dtorCalled == 1);
+    return dtorCalled;
+}
+static assert(numDtorCallsByGotoOutOfScope() == 1);
+
+
+uint numDtorCallsByGotoDifferentScopeAfter()
+{
+    uint dtorCalled;
+    {
+      S s = S(&dtorCalled);
+      assert(dtorCalled == 0);
+    }
+    assert(dtorCalled == 1);
+    goto L_abc;
+    L_abc:
+    assert(dtorCalled == 1);
+    return dtorCalled;
+}
+static assert(numDtorCallsByGotoDifferentScopeAfter() == 1);
+
+
+auto numDtorCallsByGotoDifferentScopeBefore()
+{
+    uint dtorCalled;
+    assert(dtorCalled == 0);
+    goto L_abc;
+    L_abc:
+    assert(dtorCalled == 0);
+    {
+      S s = S(&dtorCalled);
+      assert(dtorCalled == 0);
+    }
+    assert(dtorCalled == 1);
+    return dtorCalled;
+}
+static assert(numDtorCallsByGotoDifferentScopeBefore() == 1);
+
+
+struct S2
+ {
+    ~this()
+    {
+        dtorCalled = true;
+    }
+    
+    bool dtorCalled = false;
+}
+
+auto structInCaseScope()
+{
+    auto charsets = S2();
+    switch(1)
+    {
+    case 0:
+        auto set = charsets;
+        break;
+    default:
+        break;
+    }
+    return 1;
+}
+
+static assert(structInCaseScope());
```

Comment:
I investigated a few solutions to fix this bug. On the way to the final one I had one that would give a cantexp on "auto set = charsets;", regardless of the return value.
I like to err more on the side of more unit tests, so I left this one here. I can remove it though, please tell me what you think.